iceoryx_utils/test/moduletests/test_unit_duration.cpp

```diff
@@ -18,218 +18,686 @@
-    EXPECT_DOUBLE_EQ(time.microSeconds<double>(), 1.0 / KILO);
-    EXPECT_DOUBLE_EQ(time.nanoSeconds<double>(), 1.0);
+    EXPECT_EQ(time.seconds<uint64_t>(), 0U);
+    EXPECT_EQ(time.milliSeconds<uint64_t>(), 0U);
+    EXPECT_EQ(time.microSeconds<uint64_t>(), 0U);
+    EXPECT_EQ(time.nanoSeconds<uint64_t>(), 1U);
+}
+
+TEST(Duration_test, ConvertTimespecWithNoneReference)
+{
+    constexpr int64_t SECONDS{44};
+    constexpr int64_t NANOSECONDS{55};
+
+    Duration duration{SECONDS, NANOSECONDS};
+
+    struct timespec sut = timespec(duration.timespec(iox::units::TimeSpecReference::None));
+
+    EXPECT_THAT(sut.tv_sec, Eq(SECONDS));
+    EXPECT_THAT(sut.tv_nsec, Eq(NANOSECONDS));
+}
+
+TEST(Duration_test, ConvertTimespecWithMonotonicReference)
+{
+    constexpr int64_t SECONDS{4};
+    constexpr int64_t NANOSECONDS{66};
+
+    auto timeSinceUnixEpoch = std::chrono::system_clock::now().time_since_epoch();
+    auto timeSinceMonotonicEpoch = std::chrono::steady_clock::now().time_since_epoch();
+
+    Duration duration{SECONDS, NANOSECONDS};
+    struct timespec sut = timespec(duration.timespec(iox::units::TimeSpecReference::Monotonic));
+
+    auto secondsSinceUnixEpoch = std::chrono::duration_cast<std::chrono::seconds>(timeSinceUnixEpoch).count();
+    auto secondsSinceMonotonicEpoch = std::chrono::duration_cast<std::chrono::seconds>(timeSinceMonotonicEpoch).count();
+    EXPECT_THAT(sut.tv_sec, Lt(secondsSinceUnixEpoch));
+    EXPECT_THAT(sut.tv_sec, Gt(secondsSinceMonotonicEpoch));
+}
+
+TEST(Duration_test, ConvertTimespecWithEpochReference)
+{
+    constexpr int64_t SECONDS{5};
+    constexpr int64_t NANOSECONDS{77};
+
+    auto timeSinceUnixEpoch = std::chrono::system_clock::now().time_since_epoch();
+
+    Duration duration{SECONDS, NANOSECONDS};
+    struct timespec sut = timespec(duration.timespec(iox::units::TimeSpecReference::Epoch));
+
+    auto secondsSinceUnixEpoch = std::chrono::duration_cast<std::chrono::seconds>(timeSinceUnixEpoch).count();
+    EXPECT_THAT(10 * SECONDS, Lt(secondsSinceUnixEpoch));
+    EXPECT_THAT(sut.tv_sec, Gt(secondsSinceUnixEpoch));
+}
+
+// END CONVERSION FUNCTION TESTS
+
+// BEGIN CONVERSION OPERATOR TESTS
+
+TEST(Duration_test, OperatorTimeval)
+{
+    constexpr int64_t SECONDS{111};
+    constexpr int64_t MICROSECONDS{222};
+    constexpr int64_t ROUND_OFF_NANOSECONDS{666};
+
+    Duration duration{SECONDS, MICROSECONDS * KILO + ROUND_OFF_NANOSECONDS};
+
+    struct timeval sut = timeval(duration);
+
+    EXPECT_THAT(sut.tv_sec, Eq(SECONDS));
+    EXPECT_THAT(sut.tv_usec, Eq(MICROSECONDS));
 }
 
-TEST(Duration_test, comparison)
+// END CONVERSION OPERATOR TESTS
+
+// BEGIN COMPARISON TESTS
+
+TEST(Duration_test, CompareTwoEqualDurations)
 {
     auto time1 = 200_us;
-    auto time2 = 200.0_us;
-    EXPECT_EQ(time1.seconds<double>(), time2.seconds<double>());
+    auto time2 = 200000_ns;
+    EXPECT_TRUE(time1 == time2);
+    EXPECT_TRUE(time2 == time1);
+}
```

Comment:
Please add also a test where the equality operator returns false. A paranoid part of me fears that someone in the future might change the behavior of the operator so that it always returns true.